File: traceharbor-battery-canary/src/main/java/com/kernelflux/traceharbor/batterycanary/BatteryMonitorPlugin.java
package com.kernelflux.traceharbor.batterycanary;

import android.app.Application;

import com.kernelflux.traceharbor.TraceHarbor;
import com.kernelflux.traceharbor.batterycanary.monitor.BatteryMonitorConfig;
import com.kernelflux.traceharbor.batterycanary.monitor.BatteryMonitorCore;
import com.kernelflux.traceharbor.lifecycle.owners.ProcessUILifecycleOwner;
import com.kernelflux.traceharbor.plugin.Plugin;
import com.kernelflux.traceharbor.plugin.PluginListener;
import com.kernelflux.traceharbor.util.TraceHarborLog;
import com.kernelflux.traceharbor.util.TraceHarborUtil;

public class BatteryMonitorPlugin extends Plugin {
    private static final String TAG = "TraceHarbor.battery.BatteryMonitorPlugin";

    final BatteryMonitorCore mDelegate;
    private static String sPackageName = null;
    private static String sProcessName = null;

    public BatteryMonitorPlugin(BatteryMonitorConfig config) {
        mDelegate = new BatteryMonitorCore(config);
        TraceHarborLog.i(TAG, "setUp battery monitor plugin with configs: " + config);
    }

    public BatteryMonitorCore core() {
        return mDelegate;
    }

    @Override
    public void init(Application app, PluginListener listener) {
        super.init(app, listener);
        if (!mDelegate.getConfig().isBuiltinForegroundNotifyEnabled) {
            ProcessUILifecycleOwner.INSTANCE.removeListener(this);
        }
    }

    @Override
    public String getTag() {
        return "BatteryMonitorPlugin";
    }

    @Override
    public void start() {
        super.start();
        mDelegate.start();
    }

    @Override
    public void stop() {
        super.stop();
        mDelegate.stop();
    }

    @Override
    public void onForeground(boolean isForeground) {
        mDelegate.onForeground(isForeground);
    }

    @Override
    public boolean isForeground() {
        return mDelegate.isForeground();
    }

    public String getProcessName() {
        if (sProcessName == null) {
            synchronized (this) {
                if (sProcessName == null) {
                    Application application = getApplication();
                    if (application == null) {
                        if (!TraceHarbor.isInstalled()) {
                            throw new IllegalStateException(getTag() + " is not yet init!");
                        }
                        application = TraceHarbor.with().getApplication();
                    }
                    sProcessName = TraceHarborUtil.getProcessName(application);
                }
            }
        }
        return sProcessName;
    }

    public String getPackageName() {
        if (sPackageName == null) {
            synchronized (this) {
                if (sPackageName == null) {
                    Application application = getApplication();
                    if (application == null) {
                        if (!TraceHarbor.isInstalled()) {
                            throw new IllegalStateException(getTag() + " is not yet init!");
                        }
                        application = TraceHarbor.with().getApplication();
                    }
                    sPackageName = application.getPackageName();
                }
            }
        }
        return sPackageName;
    }
}

File: traceharbor-battery-canary/src/main/java/com/kernelflux/traceharbor/batterycanary/BatteryMonitorPlugin.kt
package com.kernelflux.traceharbor.batterycanary

import android.app.Application
import com.kernelflux.traceharbor.TraceHarbor
import com.kernelflux.traceharbor.batterycanary.monitor.BatteryMonitorConfig
import com.kernelflux.traceharbor.batterycanary.monitor.BatteryMonitorCore
import com.kernelflux.traceharbor.lifecycle.owners.ProcessUILifecycleOwner
import com.kernelflux.traceharbor.plugin.Plugin
import com.kernelflux.traceharbor.plugin.PluginListener
import com.kernelflux.traceharbor.util.TraceHarborLog
import com.kernelflux.traceharbor.util.TraceHarborUtil

class BatteryMonitorPlugin(config: BatteryMonitorConfig) : Plugin() {
    @JvmField
    val mDelegate: BatteryMonitorCore = BatteryMonitorCore(config)

    init {
        TraceHarborLog.i(TAG, "setUp battery monitor plugin with configs: $config")
    }

    fun core(): BatteryMonitorCore = mDelegate

    override fun init(application: Application, pluginListener: PluginListener) {
        super.init(application, pluginListener)
        if (!mDelegate.getConfig().isBuiltinForegroundNotifyEnabled) {
            ProcessUILifecycleOwner.removeListener(this)
        }
    }

    override val tag: String
        get() = "BatteryMonitorPlugin"

    override fun start() {
        super.start()
        mDelegate.start()
    }

    override fun stop() {
        super.stop()
        mDelegate.stop()
    }

    override fun onForeground(isForeground: Boolean) {
        mDelegate.onForeground(isForeground)
    }

    override fun isForeground(): Boolean = mDelegate.isForeground()


    private fun getApp(): Application {
        var app = application
        if (app == null) {
            if (!TraceHarbor.isInstalled()) {
                throw IllegalStateException("$tag is not yet init!")
            }
            app = TraceHarbor.with().application
        }
        return app
    }

    fun getProcessName(): String {
        return processName ?: synchronized(this) {
            processName ?: TraceHarborUtil.getProcessName(getApp()).also {
                processName = it
            }
        }
    }

    fun getPackageName(): String {
        return packageNameCache ?: synchronized(this) {
            packageNameCache ?: getApp().packageName.also {
                packageNameCache = it
            }
        }
    }

    private companion object {
        private const val TAG = "TraceHarbor.battery.BatteryMonitorPlugin"
        private var packageNameCache: String? = null
        private var processName: String? = null
    }
}

